katharsis-client/src/main/java/io/katharsis/client/internal/AbstractStub.java

```diff
@@ -15,120 +13,102 @@
 
 	public static final String CONTENT_TYPE = "application/vnd.api+json";
 
-	protected KatharsisClient katharsis;
+	protected KatharsisClient client;
 
 	protected JsonApiUrlBuilder urlBuilder;
 
 	public AbstractStub(KatharsisClient client, JsonApiUrlBuilder urlBuilder) {
-		this.katharsis = client;
+		this.client = client;
 		this.urlBuilder = urlBuilder;
 	}
 
-	protected BaseResponseContext executeGet(String requestUrl) {
-		return execute(requestUrl, true, HttpMethod.GET, null);
+	protected Object executeGet(String requestUrl, ResponseType responseType) {
+		return execute(requestUrl, responseType, HttpMethod.GET, null);
 	}
 
-	protected BaseResponseContext executeDelete(String requestUrl) {
-		return execute(requestUrl, false, HttpMethod.DELETE, null);
+	protected Object executeDelete(String requestUrl) {
+		return execute(requestUrl, ResponseType.NONE, HttpMethod.DELETE, null);
 	}
 
-	@SuppressWarnings("unchecked")
-	protected <T> DefaultResourceList<T> toList(JsonApiResponse response) {
-		Object entity = response.getEntity();
-		List<T> list;
-		if (entity instanceof List) {
-			list = (List<T>) entity;
-		}
-		else {
-			list = (List<T>) Arrays.asList(entity);
-		}
-		LinksInformation linksInformation = response.getLinksInformation();
-		MetaInformation metaInformation = response.getMetaInformation();
-		return new DefaultResourceList<T>(list, metaInformation, linksInformation);
+	public enum ResponseType {
+		NONE, RESOURCE, RESOURCES
 	}
 
-	protected BaseResponseContext execute(String url, boolean getResponse, HttpMethod method, String requestBody) {
+	protected Object execute(String url, ResponseType responseType, HttpMethod method, String requestBody) {
 		try {
 
-
-			HttpAdapter httpAdapter = katharsis.getHttpAdapter();
+			HttpAdapter httpAdapter = client.getHttpAdapter();
 			HttpAdapterRequest request = httpAdapter.newRequest(url, method, requestBody);
-			
+
 			LOGGER.debug("requesting {} {}", method, url);
-			if(requestBody != null){
+			if (requestBody != null) {
 				LOGGER.debug("request body: {}", requestBody);
 			}
 
 			request.header("Content-Type", CONTENT_TYPE);
 			request.header("Accept", CONTENT_TYPE);
 
 			HttpAdapterResponse response = request.execute();
+
 			if (!response.isSuccessful()) {
-				handleError(response);
+				throw handleError(response);
 			}
 
 			String body = response.body();
 			LOGGER.debug("response body: {}", body);
-			ObjectMapper objectMapper = katharsis.getObjectMapper();
-			if (getResponse) {
-				return objectMapper.readValue(body, BaseResponseContext.class);
-			}
-			else {
+			ObjectMapper objectMapper = client.getObjectMapper();
+
+			if (responseType != ResponseType.NONE) {
+				Document document = objectMapper.readValue(body, Document.class);
+
+				ClientDocumentMapper documentMapper = client.getDocumentMapper();
+				return documentMapper.fromDocument(document, responseType == ResponseType.RESOURCES);
+			} else {
```

Comment:
You can remove this else statement and just put `return null` at the end of the function - thus reducing complexibty